hal/src/gcc/exflash_hal.cpp

```diff
@@ -0,0 +1,210 @@
+
+    void erase(uintptr_t addr, size_t blockCount, size_t blockSize) {
+        if (!blockSize) {
+            return;
+        }
+        addr = addr / blockSize * blockSize;
+        size_t size = blockSize * blockCount;
+        if (addr + size > EXTERNAL_FLASH_SIZE) {
+            throw std::runtime_error("Invalid address");
+        }
+        std::lock_guard lock(mutex_);
+        buf_.erase(addr, size);
+        bufferChanged();
+    }
+
+    void lock() {
+        mutex_.lock();
+    }
+
+    void unlock() {
+        mutex_.lock();
+    }
+
+    static ExternalFlash* instance() {
+        static ExternalFlash f;
+        return &f;
+    }
+
+private:
+    SparseBuffer buf_;
+    std::string tempFile_;
+    std::string persistFile_;
+
+    mutable std::recursive_mutex mutex_;
+
+    ExternalFlash() :
+            buf_(0xff /* fill */) {
+        if (!deviceConfig.flash_file.empty()) {
+            persistFile_ = deviceConfig.flash_file;
+            if (fs::exists(persistFile_)) {
+                loadBuffer(buf_, persistFile_);
+            }
+            fs::path p(persistFile_);
+            tempFile_ = p.parent_path().append('~' + p.filename().string());
+        }
+    }
+
+    void bufferChanged() {
+        if (!persistFile_.empty()) {
+            saveBuffer(buf_, tempFile_);
+            fs::rename(tempFile_, persistFile_);
+        }
+    }
+
+    static void loadBuffer(SparseBuffer& buf, const std::string& file) {
+        std::ifstream f;
+        f.exceptions(std::ios::badbit | std::ios::failbit);
+        f.open(file, std::ios::binary);
+        size_t segCount = readUint32(f);
+        for (size_t i = 0; i < segCount; ++i) {
+            size_t offs = readUint32(f);
+            size_t size = readUint32(f);
+            std::string s;
+            s.resize(size);
+            f.read(s.data(), size);
+            buf.write(offs, s);
+        }
+    }
+
+    static void saveBuffer(const SparseBuffer& buf, const std::string& file) {
+        std::ofstream f;
+        f.exceptions(std::ios::badbit | std::ios::failbit);
+        f.open(file, std::ios::binary | std::ios::trunc);
+        auto& seg = buf.segments();
+        writeUint32(f, seg.size());
+        for (auto it = seg.begin(); it != seg.end(); ++it) {
+            writeUint32(f, it->first);
+            writeUint32(f, it->second.size());
+            f.write(it->second.data(), it->second.size());
+        }
+        f.close();
+    }
+
+    static uint32_t readUint32(std::ifstream& f) {
+        uint32_t v = 0;
+        f.read((char*)&v, sizeof(v));
+        return littleEndianToNative(v);
+    }
+
+    static void writeUint32(std::ofstream& f, uint32_t val) {
+        val = nativeToLittleEndian(val);
+        f.write((const char*)&val, sizeof(val));
+    }
+};
+
+} // namespace
+
+int hal_exflash_init(void) {
+    return 0;
+}
+
+int hal_exflash_uninit(void) {
+    return 0;
+}
+
+int hal_exflash_read(uintptr_t addr, uint8_t* data_buf, size_t data_size) {
+    try {
```

Comment:
While this pattern is not used in normal Device OS sources for real devices, would be nice to introduce some CHECK-style macro to forward `SYSTEM_ERROR_XXX` error codes on exceptions to improve readability.